R/checkers.R

```diff
@@ -45,3 +56,213 @@ check_elements_non_neg <- function(x, arg = "x", call = rlang::caller_env()) {
+    throw_type_error(
+      object = x,
+      arg_name = arg,
+      expected_type = "integer or character",
+      call = call
+    )
+  }
+  invisible()
+}
+
+
+#' Throw an informative type error on a user-provided input
+#'
+#' Follows the guidance from [rlang::abort()] on applying a call and a class
+#' in the error message. Used as a base in type-checkers to throw a properly
+#' formatted error.
+#'
+#' @param object Object with incorrect type
+#' @param arg_name Name of the argument corresponding to `object`
+#' @param expected_type The type that the user should provide instead
+#' @param call The calling environment to be reflected in the error message
+#'
+#' @return This function is called for its side effect of throwing an error. It
+#'   should never return.
+throw_type_error <- function(object,
+                             arg_name,
+                             expected_type,
+                             call = rlang::caller_env()) {
+  cli::cli_abort(
+    c("{.arg {arg_name}} is {.obj_type_friendly {object}}",
+      "i" = "Must be of type {.emph {expected_type}}"
+    ),
+    call = call,
+    class = "wwinference_type_error"
+  )
+}
+
+
+#' Check that the input wastewater data contains all the required column names
+#'
+#' @description
+#' This function is intended to be used to check that the wastewater data that
+#' gets passed into [preprocess_ww_data()] contains the required columns. If
+#' it does not, we want to tell the user which column are missing. This will not
+#' however, ensure that the elements of the column are of the right type,
+#' or check that the values of them make sense.
+#'
+#'
+#' @param ww_data tibble containing the input wastewater data
+#' @param conc_col_name string indicating the name of the column containing
+#' the concentration measurements in the wastewater data
+#' @param lod_col_name string indicating the name of the column containing
+#' the concentration measurements in the wastewater data
+#' @param add_req_col_names vector of strings indicating the required wastewater
+#' column names, the defaults are `c("date", "site", "lab", "site_pop")`
+#' @param call Calling environment to be passed to [cli::cli_abort()] for
+#' traceback.
+#'
+#' @return NULL, invisibly
+check_required_ww_inputs <- function(ww_data,
+                                     conc_col_name,
+                                     lod_col_name,
+                                     add_req_col_names = c(
+                                       "date", "site",
+                                       "lab", "site_pop"
+                                     ),
+                                     call = rlang::caller_env()) {
+  column_names <- colnames(ww_data)
+  expected_col_names <- c(
+    {
+      conc_col_name
+    },
+    {
+      lod_col_name
+    },
+    add_req_col_names
+  )
+
+  # This tells you whats missing
+  checkmate::check_names(column_names, must.include = expected_col_names)
```

Comment:
@dylanhmorris I think this would be where I would want to maybe submit an issue to `checkmate` because we don't want to have to assert this twice -- the first gives you a useful error message about what names are missing, but I also want a custom error message that puts it in context, e.g. what data are we applying this to (which I have added below).

Reply:
https://github.com/CDCgov/ww-inference-model/pull/37/files#r1702288282

Reply:
I do agree that it would be nice to be able to customize the error message directly in `assert_`.